AdvancedWidget: Replace FFV1 codec with Ut Video

TryTwo opened the pull request:
Replaces FFV1 codec checkbox with Ut Video.  To still use FFV1, one would have to uncheck "Use lossless" and set DumpCodec =  ffv1, which is currently what it takes to use Ut Video.

Users who previously had FFV1 checked will need to re-enable the Use Lossless replacement.

Requested here:
https://bugs.dolphin-emu.org/issues/13657

<details>
  <summary>Abandoned Plan</summary>

![codec](https://github.com/user-attachments/assets/8675724d-11c4-49c9-b6d4-7b53c6085cd6)

Requested here:
https://bugs.dolphin-emu.org/issues/13657

Problem: "UseFFV1" should no longer be needed, but I'm not sure how/where to convert that over so it doesn't collide.

</details>

vadosnaprimer commented on Source/Core/DolphinQt/Config/Graphics/AdvancedWidget.cpp:
```diff
@@ -185,13 +185,23 @@ void AdvancedWidget::CreateWidgets()
+  // Update EnableBitrate() if options are changed.
+
+  const std::vector<std::pair<QString, QString>> options{
+      {tr("Auto"), QStringLiteral("")},
```
What does the `Auto` option do? It's not explained in the popup, and I personally don't think it's needed at all.

TryTwo replied:
It sets nothing, which is currently how it works. I'm not sure of how every OS handles that. It may be safe to not have Auto.

vadosnaprimer replied:
If it's set to nothing then what happens when one tries to dump frames on `Auto`?

vadosnaprimer commented on Source/Core/DolphinQt/Config/Graphics/AdvancedWidget.cpp:
```diff
@@ -391,9 +414,11 @@ void AdvancedWidget::AddDescriptions()
-                 "unsure, leave this unchecked.</dolphin_emphasis>");
+  static const char TR_CODEC_CHOICE_DESCRIPTION[] = QT_TR_NOOP(
+      "Selects codec for encoding frame dumps. <br>MPEG4: Lossy. Good compression. Can set "
+      "bitrate.<br>FFV1: Lossless. Good compression. Slow.<br>UTVideo: Lossless. Fast. Poor "
```
```suggestion
      "bitrate.<br>FFV1: Lossless. Good compression. Slow.<br>UTVideo: Lossless. Fast. Low "
```

vadosnaprimer commented on Source/Core/DolphinQt/Config/Graphics/AdvancedWidget.cpp:
```diff
@@ -391,9 +414,11 @@ void AdvancedWidget::AddDescriptions()
+  static const char TR_CODEC_CHOICE_DESCRIPTION[] = QT_TR_NOOP(
+      "Selects codec for encoding frame dumps. <br>MPEG4: Lossy. Good compression. Can set "
+      "bitrate.<br>FFV1: Lossless. Good compression. Slow.<br>UTVideo: Lossless. Fast. Poor "
+      "compression that produces large video files. Good compatibility with editing "
```
```suggestion
      "compression. High compatibility with editing "
```

TryTwo replied:
I'd like to mention somewhere that Ut Video makes large files, to be double sure the user is aware.

vadosnaprimer commented:
I recommend putting FFV1 last due to how slow and incompatible it is.

vadosnaprimer commented on Source/Core/DolphinQt/Config/Graphics/AdvancedWidget.cpp:
```diff
@@ -391,9 +414,11 @@ void AdvancedWidget::AddDescriptions()
-      QT_TR_NOOP("Encodes frame dumps using the FFV1 codec.<br><br><dolphin_emphasis>If "
-                 "unsure, leave this unchecked.</dolphin_emphasis>");
+  static const char TR_CODEC_CHOICE_DESCRIPTION[] = QT_TR_NOOP(
+      "Selects codec for encoding frame dumps. <br>MPEG4: Lossy. Good compression. Can set "
```
```suggestion
      "Selects codec for encoding frame dumps.
      <br><br><b>MPEG4</b>: Lossy. Good compression. Can set "
```
For format consistency with other popups.

vadosnaprimer commented on Source/Core/DolphinQt/Config/Graphics/AdvancedWidget.cpp:
```diff
@@ -391,9 +414,11 @@ void AdvancedWidget::AddDescriptions()
+      "Selects codec for encoding frame dumps. <br>MPEG4: Lossy. Good compression. Can set "
+      "bitrate.<br>FFV1: Lossless. Good compression. Slow.<br>UTVideo: Lossless. Fast. Poor "
+      "compression that produces large video files. Good compatibility with editing "
+      "programs.<br><br><dolphin_emphasis>If unsure, leave this unchecked.</dolphin_emphasis>");
```
```suggestion
      "programs."
      "<br><br><dolphin_emphasis>If unsure, leave this unchecked.</dolphin_emphasis>");
```

vadosnaprimer commented on Source/Core/DolphinQt/Config/Graphics/AdvancedWidget.cpp:
```diff
@@ -185,13 +185,23 @@ void AdvancedWidget::CreateWidgets()
+      {tr("Auto"), QStringLiteral("")},
+      {QStringLiteral("MPEG4"), QStringLiteral("mpeg4")},
+      {QStringLiteral("FFV1"), QStringLiteral("ffv1")},
+      {QStringLiteral("UTVideo"), QStringLiteral("utvideo")}};
```
Legend (from mediainfo:

```
Format                      : MPEG-4 Visual
Format settings, Matrix     : Default (H.263)
Codec ID                    : XVID
Codec ID/Hint               : XviD
```

```
Format                      : RGB
Codec ID                    : ULRG
Codec ID/Info               : Ut Video Lossless Codec
Codec ID/Hint               : Ut Video
```

```
Format                      : FFV1
Codec ID                    : FFV1
```

I think it's better to name the options after codec names, so `MPEG4` (which means format) would become `XviD`, and `UTVideo` - `Ut Video`.

TryTwo commented:
I'm not certain about using High./Low instead of Good/Poor/Bad.  I just want it to be clear which is better, especially when considering translations.  Maybe it's a non-issue.

I added more code to toggle off the old FFV1 bool option.  It will only trigger when opening the options though.

vadosnaprimer commented:
"Poor compression" to me means "really bad compression, little to none, it does poor job at it", which is far from the case for UT. It's just not as heavy as FFV1 or something like Lagarith. Which is compensated by high speed etc.

Lossy is better for regular usage, UT is better for high quality encoding (it's reencoded to lossy on our end afterwards).

vadosnaprimer commented on Source/Core/DolphinQt/Config/Graphics/AdvancedWidget.cpp:
```diff
@@ -391,9 +427,13 @@ void AdvancedWidget::AddDescriptions()
-                 "unsure, leave this unchecked.</dolphin_emphasis>");
+  static const char TR_CODEC_CHOICE_DESCRIPTION[] =
+      QT_TR_NOOP("Selects codec for encoding frame dumps."
+                 "<br><b>Xvid</b>: Lossy. Good compression. Can set bitrate."
```
The above popup for resolution uses double linebreak between the meaning and the options.

TryTwo replied:
Oops, thanks

TryTwo commented:
What if we drop mentioning compression and just say Ut Video produces larger files?

vadosnaprimer commented:
I think if it's changed into `Lossless. Fast. Low compression (large file size). Good compatibility with editing programs.` then it will be perfect.

mbc07 commented:
What are the use cases for FFV1, if any? Considering the difficulties in converting/editing the video after the dump, maybe drop FFV1 in favour of UTVideo? Seems a bit redundant to have multiple lossless codecs as choices...

vadosnaprimer commented:
> What are the use cases for FFV1, if any? Considering the difficulties in converting/editing the video after the dump, maybe drop FFV1 in favour of UTVideo? Seems a bit redundant to have multiple lossless codecs as choices...

Honestly I agree. Maybe it's time to finally let it go indeed. I don't know of a single use case for FFV1 anymore where UT is unfeasible.

TryTwo commented:
It'd be easy to just change the FFV1 checkbox into a Ut Video checkbox (or just call it lossless). I don't know if it's okay to just assume nobody uses a FFV1 setup though, as it produces smaller files.

Are there any other options that should be added? I'm guessing xvid is enough for lossy. Is there another, better lossless option? It looks like h.264 with -crf0 is lossless? Then you can change the compression rate vs speed. Or is that not compatible with TASing?

mbc07 commented:
Isn't XVID the same situation as FFV1 regarding VFW support? I don't know if the FFmpeg build we compile currently allows this as AFAICT it can quickly get the final binary very bloated if we enable all features it has, but in the GUI I'd offer one lossy option, one lossless option, and that's it. There's already an INI-only setting if more tech-savvy users want to select a specific codec, and I think it's enough (we could expose that directly in the GUI if there's enough demand, though, the existing codec options lives in the "Advanced" tab after all).

Stick with either H.264 or H.265 for lossy encoding (or perhaps VP9 if we want a royalty-free codec), since any of those are supported practically everywhere nowadays, and additionally offer a lossless option (FFV1 seems bad, so perhaps UTvideo, or other codec?)

TryTwo commented:
Also, I don't actually know how to do anything besides set the string for which codec to use.

vadosnaprimer commented:
> I don't know if it's okay to just assume nobody uses a FFV1 setup though, as it produces smaller files.

Back in the day when the exact format of the FFV1 option was barely usable, I only know of the TAS people asking to improve its compatibility. And they (we) already largely switched to UT because decoding FFV1 is such a PITA still. It's small but it's **very** slow during encoding, and even at seeking (important when editing the video). The assumption is not that nobody uses FFV1 anymore but that nobody is going to specifically need it when UT is there.

> It looks like h.264 with -crf0 is lossless? Then you can change the compression rate vs speed. Or is that not compatible with TASing?

x264 is very fast and compresses very well, but its seeking has big problems in VFW based video editors (like avisynth). The codec got some updates over the years that import plugins for it didn't have, so with ffmpeg's libx264 seeking is outright broken. UT's seeking is instant even for giant 4K dumps.

> Isn't XVID the same situation as FFV1 regarding VFW support?

I haven't specifically tested it but I think XVID is much more widely supported by decoders.

> I don't know if the FFmpeg build we compile currently allows this as AFAICT it can quickly get the final binary very bloated if we enable all features it has, but in the GUI I'd offer one lossy option, one lossless option, and that's it.

Personally I'd only enable libx264 and utvideo (and let the user send string arguments to them, ideally via nut container but that seems to involve overhauling the whole dumping system). h264 is super well supported and has crazy amount of options, and likely compresses much more efficiently than xvid.

> There's already an INI-only setting if more tech-savvy users want to select a specific codec

I don't know if this works, because of how few things are included in dolphin's ffmpeg.

> Stick with either H.264 or H.265 for lossy encoding (or perhaps VP9 if we want a royalty-free codec), since any of those are supported practically everywhere nowadays

h264's compatibility is still higher, and h265/vp9 don't offer dramatic improvements if we care about speed versus quality. Modern codecs are more efficient but also quite a bit slower than x264, so if you balance everything out they all have about the same sweet spot at speed/compression/quality, so there's little benefit for regular users.

TryTwo commented:
Thanks for all this info.

So, do we want to drop the combobox and just change the FFV1 checkbox to UT Video?  Any future plans can have two main options lossy/lossless, and then maybe more extra config options.

If so, should it be called  "Ut Video" or "Lossless"? I had never heard of Ut Video before, so as a User who knows little Lossless is more understandable.

If not, then I think this PR should just keep the combobox how it is, and someone else could mess with it more later.

vadosnaprimer commented:
I'd call the 2 options `Xvid lossy` and `UtVideo lossless`.

TryTwo commented:
I turned it back into a checkbox, for utvideo.  I just removed UseFFV1 and didn't do anything with it.  If someone wants to use ffv1, with how it's coded, they would have to set the dump string to "ffv1" and not enable Use Lossless (to avoid an override).

The option will be off by default for everyone, they'd have to enable it even if ffv1 was enabled before.  I think that's okay because it indicates that the option has changed?

I updated the description a little.

If this is the wrong approach let me know.

mbc07 reviewed (approved):
LGTM. I have no particular preference in regards to migrating the old setting to the new one in this case, and I think having the new INI key be just "UseLossless" future proofs any further (lossless) codec change. The updated label/description is also fine...

TryTwo commented:
@JMC47  Do you have any thoughts on how this change is being done?
